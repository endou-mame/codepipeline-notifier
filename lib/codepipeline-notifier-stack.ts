import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as sns from 'aws-cdk-lib/aws-sns';
import { type Construct } from 'constructs';

type CodepipelineNotifierStackProps = {
  teamsWebhookUrl: string;
} & cdk.StackProps;

export class CodepipelineNotifierStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CodepipelineNotifierStackProps) {
    super(scope, id, props);

    const codepipelineNotifierForTeams = new NodejsFunction(
      this,
      'CodepipelineNotifierForTeamsLambda',
      {
        functionName: 'codepipeline-notifier-for-teams',
        entry: 'src/for-teams/index.ts',
        handler: 'handler',
        environment: {
          TEAMS_WEBHOOK_URL: props.teamsWebhookUrl,
        },
        runtime: lambda.Runtime.NODEJS_20_X,
      },
    );

    const topic = new sns.Topic(this, 'CodepipelineNotifierSnsTopic', {
      displayName: 'codepipeline-notifier',
      topicName: 'codepipeline-notifier',
    });

    codepipelineNotifierForTeams.addEventSource(new lambdaEventSources.SnsEventSource(topic));
  }
}
