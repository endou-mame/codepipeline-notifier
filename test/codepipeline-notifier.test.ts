import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import cdkJson from '../cdk.json';
import { CodepipelineNotifierStack } from '../lib/codepipeline-notifier-stack';

describe('CodepipelineNotifierStack Assertion Test', () => {
  let template: Template;

  // When
  beforeAll(() => {
    /**
     * FIXME: https://github.com/aws/aws-cdk/issues/18125
     */
    const context = {
      ...cdkJson.context,
      'aws:cdk:bundling-stacks': [],
    };
    const app = new cdk.App({ context });
    const stack = new CodepipelineNotifierStack(app, 'CodepipelineNotifierStack', {
      teamsWebhookUrl: 'dummy',
    });
    template = Template.fromStack(stack);
  });

  test('SNS Topic: リソースが1つ作成されている', () => {
    template.resourceCountIs('AWS::SNS::Topic', 1);
  });
  test('SNS Topic: プロパティが正しい', () => {
    template.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: 'codepipeline-notifier',
    });
  });

  // Then
  test('Lambda: リソースが1つ作成されている', () => {
    template.resourceCountIs('AWS::Lambda::Function', 1);
  });
  test('Lambda: プロパティが正しい', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      FunctionName: 'codepipeline-notifier-for-teams',
    });
  });
});

describe('CodepipelineNotifierStack Snapshot Test', () => {
  test('Snapshot が一致すること', () => {
    const app = new cdk.App();
    const stack = new CodepipelineNotifierStack(app, 'CodepipelineNotifierStack', {
      teamsWebhookUrl: 'dummy',
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
