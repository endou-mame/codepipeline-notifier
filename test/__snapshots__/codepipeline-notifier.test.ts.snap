// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodepipelineNotifierStack Snapshot Test Snapshot が一致すること 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CodepipelineNotifierForTeamsLambda4B2B5B50": {
      "DependsOn": [
        "CodepipelineNotifierForTeamsLambdaServiceRole5C8F4E3A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5ea48f7de447a7bc5b818d257ea8dac52b58ea0f09f60f48c3339797a70c9592.zip",
        },
        "Environment": {
          "Variables": {
            "TEAMS_WEBHOOK_URL": "dummy",
          },
        },
        "FunctionName": "codepipeline-notifier-for-teams",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CodepipelineNotifierForTeamsLambdaServiceRole5C8F4E3A",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CodepipelineNotifierForTeamsLambdaAllowInvokeCodepipelineNotifierStackCodepipelineNotifierSnsTopic065B5B5C34E95183": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CodepipelineNotifierForTeamsLambda4B2B5B50",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "CodepipelineNotifierSnsTopicC453A45B",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CodepipelineNotifierForTeamsLambdaCodepipelineNotifierSnsTopic9F3F1ADB": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "CodepipelineNotifierForTeamsLambda4B2B5B50",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "CodepipelineNotifierSnsTopicC453A45B",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "CodepipelineNotifierForTeamsLambdaServiceRole5C8F4E3A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CodepipelineNotifierSnsTopicC453A45B": {
      "Properties": {
        "DisplayName": "codepipeline-notifier",
        "TopicName": "codepipeline-notifier",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
